src/libxively/event_dispatcher/xi_time_event.c

```diff
@@ -0,0 +1,426 @@
+/* Copyright (c) 2003-2016, LogMeIn, Inc. All rights reserved.
+ *
+ * This is part of the Xively C Client library,
+ * it is licensed under the BSD 3-Clause license.
+ */
+
+#include "xi_time_event.h"
+
+/**
+ * @brief This part of the file implements time event functionality. This
+ * implementation assumes that the element type is always the xi_time_event_t.
+ *
+ * It uses the vector as a container type. The vector stores pointers to the time events.
+ * Time events are always sorted in the vector by the time event execution time.
+ */
+
+/*
+ * STATIC INTERNAL FUNCTIONS
+ */
+
+/**
+ * @brief xi_swap_time_events
+ *
+ * Swaps the two xi_time_event_t elements in the provided vector.
+ *
+ *
+ * @param vector
+ * @param lhs_index
+ * @param rhs_index
+ */
+static void xi_swap_time_events( xi_vector_t* vector,
+                                 xi_vector_index_type_t lhs_index,
+                                 xi_vector_index_type_t rhs_index )
+{
+    /* PRE-CONDITIONS */
+    assert( NULL != vector );
+    assert( lhs_index < vector->elem_no );
+    assert( rhs_index < vector->elem_no );
+
+    xi_time_event_t* lhs_time_event =
+        ( xi_time_event_t* )vector->array[lhs_index].selector_t.ptr_value;
+    xi_time_event_t* rhs_time_event =
+        ( xi_time_event_t* )vector->array[rhs_index].selector_t.ptr_value;
+
+    xi_vector_swap_elems( vector, lhs_index, rhs_index );
+
+    lhs_time_event->position = rhs_index;
+    rhs_time_event->position = lhs_index;
+}
+
+/**
+ * @brief xi_time_event_bubble_core
+ *
+ * Generic bubbler. This function eliminates code repeatition for the logic of bubbling
+ * an element in both directions in the vector. It is used by
+ * xi_time_event_bubble_and_sort_down and xi_time_event_bubble_and_sort_up.
+ *
+ * @param vector - input vector, container of time events
+ * @param lhs - value that will appear on the left hand side of the stop condition
+ * @param rhs - value that will appear on the right hand side of the stop condition
+ * @param elem_index_var_pos - points which: lhs or rhs represents the index of an element
+ * @param dir - direction of bubbling - +1 stands for up -1 means down
+ * @return new index of bubbled element
+ */
+static xi_vector_index_type_t xi_time_event_bubble_core( xi_vector_t* vector,
+                                                         xi_vector_index_type_t lhs,
+                                                         xi_vector_index_type_t rhs,
+                                                         uint8_t elem_index_var_pos,
+                                                         int8_t dir )
+{
+    xi_vector_index_type_t* indexes[] = {&lhs, &rhs};
+
+    while ( lhs < rhs )
```

Comment:
can't we get rid of the attribute elem_index_var_pos and the local variable indexes with this solution?
while ( lhs < rhs )
    {
        const xi_vector_index_type_t *elem_index = ( 0 < dir ) ? &lhs : &rhs;

        const xi_time_event_t* lhs_element =
            ( xi_time_event_t* )vector->array[dir < 0 ? *elem_index : *elem_index + dir]
                .selector_t.ptr_value;
        const xi_time_event_t* rhs_element =
            ( xi_time_event_t* )vector->array[dir < 0 ? *elem_index + dir : *elem_index]
                .selector_t.ptr_value;

        if ( lhs_element->time_of_execution < rhs_element->time_of_execution )
        {
            /* if the elements are not in the right order lets swap them */
            xi_swap_time_events( vector, lhs_element->position, rhs_element->position );

            *elem_index += dir;
        }
        else
        {
            /* thanks to the container invariant on the order of the elemements we can
             * break at this moment */
            break;
        }
    }

    return *elem_index;